Add Rez

cooljeanius opened the pull request:
## Description
See issue #6107 

## Checklist:
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      - https://github.com/search?q=path%3A*.r+DITL&type=code
      - https://github.com/search?q=path%3A*.r+verUS&type=code
      - https://github.com/search?q=path%3A*.r+purgeable&type=code
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/cy384/ssheven/blob/master/ssheven-icons.r
      - https://github.com/cy384/ssheven/blob/master/ssheven.r
      - https://github.com/chrisballinger/rezilla/blob/72c4cfd1a2d45565d2b429bd6afff8bf426f5c65/RezillaSource/Rezilla_Rsrc/Rezilla_DITL.r
      - https://github.com/autc04/Retro68/blob/5f882506013a0a8a4335350197a1b7c91763494e/Samples/Dialog/dialog.r
      - https://github.com/vim/vim-history/blob/868c2742154cbb6d59bfd2818724b980c50eb24b/src/gui_mac.r
    - Sample license(s):
      - https://github.com/cy384/ssheven/blob/master/LICENSE
      - https://github.com/autc04/Retro68/blob/master/COPYING
      - https://github.com/vim/vim/blob/master/LICENSE
  - [x] I have included a syntax highlighting grammar: https://github.com/textmate/rez.tmbundle
  - [x] I have added a color
    - Hex value: `#FFDAB3`
    - Rationale: It's the most commonly-used color in the ResEdit icon that isn't confusable with R (see comment from @uliwitness below)
  - [x] I have updated the heuristics to distinguish my language from others using the same extension. (WIP)

uliwitness commented:
If you're looking for a color: R has blue and grey, so I'd say pick one of the remaining colors from the ResEdit icon, which seem to be green, red, or a light salmon ("flesh") color.

cooljeanius commented:
> If you're looking for a color: R has blue and grey, so I'd say pick one of the remaining colors from the ResEdit icon, which seem to be green (00CE00), red (E60100), or a light salmon ("flesh") color (FFDAB3).

OK, I counted how many pixels of each there were in the icon:
- green (both 00CE00 and a slightly different shade): 7
- E60100: 24
- FFDAB3: 44

This suggests to me that we should go with the last of the 3.

> For detecting, the most common resource types are probably 'vers', 'SIZE', 'ICON', 'cicn', 'ICN#', 'icns', 'BNDL', 'DITL', 'DLOG', 'ALRT', 'MENU', 'CURS' and 'PICT'. Might also be worth looking for `resource '****'` (with **** being a pattern, 4 characters) and `data '****'`, and maybe `#include <Types.r>` and `#include <Carbon/Carbon.r>`.

OK, could I get some help with a regex to catch all of those? I'm not very good with regexes...

uliwitness commented:
Actually, this might be a better heuristic:
```
(resource|data|type)\s+'[A-Za-z0-9]{4}'\s+((\(.*\)\s){0,1})+{
```
That will basically catch the first line of a resource declaration, as long as they don't have extra line breaks.

uliwitness commented:
For detecting some common types, one would do
```
'(ICON|BNDL|vers|SIZE|cicn|ICN#|icns|DITL|DLOG|ALRT|MENU|MBAR|CURS|PICT|cfrg|thng)'
```

uliwitness commented:
To detect the includes one would do
```
#include\s+["<](Types\.r|Carbon/Carbon\.r)[">]
```
I guess detecting any of these three regexes could be seen to indicate that it's a Rez file. So maybe something like:
```
(#include\s+["<](Types\.r|Carbon/Carbon\.r)[">])|('(ICON|BNDL|vers|SIZE|cicn|ICN#|icns|DITL|DLOG|ALRT|MENU|MBAR|CURS|PICT|cfrg|thng)')|((resource|data|type)\s+'[A-Za-z0-9]{4}'\s+((\(.*\)\s){0,1})+{)
```
should probably catch something in most files. Though some of these might give false positives if the files don't have the '.r' suffix. That said, I haven't seen an R source file yet, so we might wanna check that none of these are valid/common in that other language.

cooljeanius commented:
> To detect the includes one would do
> 
> ```
> #include\s+["<](Types\.r|Carbon/Carbon\.r)[">]
> ```
> 
> I guess detecting any of these three regexes could be seen to indicate that it's a Rez file. So maybe something like:
> 
> ```
> (#include\s+["<](Types\.r|Carbon/Carbon\.r)[">])|('(ICON|BNDL|vers|SIZE|cicn|ICN#|icns|DITL|DLOG|ALRT|MENU|MBAR|CURS|PICT|cfrg|thng)')|((resource|data|type)\s+'[A-Za-z0-9]{4}'\s+((\(.*\)\s){0,1})+{)
> ```
> 
> should probably catch something in most files. Though some of these might give false positives if the files don't have the '.r' suffix. That said, I haven't seen an R source file yet, so we might wanna check that none of these are valid/common in that other language.

There are some examples of R sources in linguist's samples directory, if you want to take a look

uliwitness commented:
OK, the `import.r` sample file in there has single-quoted strings like `'name'`. While I haven't seen `'ICON'` or the like yet, we should probably remove that part from the regex and just look for actual syntax, conservatively:

```
(#include\s+["<](Types\.r|Carbon/Carbon\.r)[">])|((resource|data|type)\s+'[A-Za-z0-9]{4}'\s+((\(.*\)\s){0,1})+{)
```

uliwitness commented:
Oh, I also think I overlooked how one specifies a resource name in Rez.

cooljeanius commented:
> OK, the `import.r` sample file in there has single-quoted strings like `'name'`. While I haven't seen `'ICON'` or the like yet, we should probably remove that part from the regex and just look for actual syntax, conservatively:
> 
> ```
> (#include\s+["<](Types\.r|Carbon/Carbon\.r)[">])|((resource|data|type)\s+'[A-Za-z0-9]{4}'\s+((\(.*\)\s){0,1})+{)
> ```

ok, so does that need to be quoted, or can it just be pasted as-is? Generally the other regexes in `heuristics.yml` are surrounded by single-quotes, and I see single-quotes in that string, so I wonder if they'll need to be escaped?

uliwitness commented:
Oh yeah since that uses YAML, you'll prolly have to escape the backslashes and the single quotes:

```
(#include\\s+["<](Types\\.r|Carbon/Carbon\\.r)[">])|((resource|data|type)\\s+\'[A-Za-z0-9]{4}\'\\s+((\\(.*\\)\\s+){0,1}){)
```

cooljeanius commented:
> Oh yeah since that uses YAML and that single-quotes these regexes, you'll prolly have to escape the backslashes and the single quotes, good catch. So the entire, quoted regex should be something like:
> 
> ```
> '(#include\\s+["<](Types\\.r|Carbon/Carbon\\.r)[">])|((resource|data|type)\\s+\'[A-Za-z0-9]{4}\'\\s+((\\(.*\\)\\s+){0,1}){)'
> ```

OK thanks, I made this the regex in 2837cbc. Unfortunately bundler is giving me some issues that are preventing me from testing locally, so hopefully a maintainer will let the CI workflows run...

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -594,6 +594,8 @@ disambiguations:
   - language: R
     pattern: '<-|^\s*#'
+  - language: Rez
+    pattern: '(#include\\s+["<](Types\\.r|Carbon/Carbon\\.r)[">])|((resource|data|type)\\s+\'[A-Za-z0-9]{4}\'\\s+((\\(.*\\)\\s+){0,1}){)'
```
There's no need to escape the backslashes

lildude commented on lib/linguist/languages.yml:
```diff
@@ -5886,6 +5886,15 @@ RenderScript:
+  type: programming
+  extensions:
+  - ".r"
+  - ".rez"
```
We need samples and search links for all extensions being added. Keep in mind our documented usage requirements so it might be better to hold off on adding a rarely used extension on the first attempt as an override can be used until usage meets requirements.

cooljeanius replied:
ok so I'll just remove the `.rez` one, then

cooljeanius replied:
removed in [7445872](https://github.com/github-linguist/linguist/pull/6524/commits/744587214c6ebda7f4eada0f9c8c1c27db0ac487).

lildude reviewed (changes requested):
See comments inline.

Please remove or replace the two large samples that are currently collapsed in the PR view; they're too big with the icons file mostly filled stuff that isn't relevant to the language.

We also need the cached license file the `script/add-grammar` should have downloaded when your added the grammar.

cooljeanius commented:
> See comments inline.
>

OK, addressed.

> Please remove or replace the two large samples that are currently collapsed in the PR view; they're too big with the icons file mostly filled stuff that isn't relevant to the language.

Removed in [769cf15](https://github.com/github-linguist/linguist/pull/6524/commits/769cf15eb54e1405bacff8e58a1f8543013adb34).

> We also need the cached license file the `script/add-grammar` should have downloaded when your added the grammar.

Hm, I guess I should just re-run the script then, if I'm not seeing it?

cooljeanius commented:
> > We also need the cached license file the `script/add-grammar` should have downloaded when your added the grammar.
> 
> Hm, I guess I should just re-run the script then, if I'm not seeing it?

OK, re-ran; is the file added in [dadde18](https://github.com/github-linguist/linguist/pull/6524/commits/dadde186731db084f616583442ab663d9952f1ed) what you were requesting?

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -594,6 +594,8 @@ disambiguations:
   - language: R
     pattern: '<-|^\s*#'
+  - language: Rez
+    pattern: '(#include\s+["<](Types\.r|Carbon/Carbon\.r)[">])|((resource|data|type)\s+\'[A-Za-z0-9]{4}\'\s+((\(.*\)\s+){0,1}){)'
```
You've got several problems here:

1. The `/` needs to be escaped
2. The `'` chars need to be escaped in the YAML way. No need for escaping in Ruby.

```suggestion
    pattern: '(#include\s+["<](Types\.r|Carbon\/Carbon\.r)[">])|((resource|data|type)\s+''[A-Za-z0-9]{4}''\s+((\(.*\)\s+){0,1}){)'
```

lildude reviewed (changes requested):
> OK, re-ran; is the file added in [dadde18](https://github.com/github-linguist/linguist/pull/6524/commits/dadde186731db084f616583442ab663d9952f1ed) what you were requesting?

Yup. That's the one.

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -594,6 +594,8 @@ disambiguations:
     pattern: '(?i:\bRebol\b)'
   - language: R
     pattern: '<-|^\s*#'
+  - language: Rez
+    pattern: '(#include\s+["<](Types\.r|Carbon\/Carbon\.r)[">])|((resource|data|type)\s+''[A-Za-z0-9]{4}''\s+((\(.*\)\s+){0,1}){)'
```
Swapping these around as R is very generic compared to Rez:

```suggestion
  - language: Rez
    pattern: '(#include\s+["<](Types\.r|Carbon\/Carbon\.r)[">])|((resource|data|type)\s+''[A-Za-z0-9]{4}''\s+((\(.*\)\s+){0,1}){)'
  - language: R
    pattern: '<-|^\s*#'
```

lildude commented on vendor/licenses/git_submodule/rez.tmbundle.dep.yml:
```diff
@@ -0,0 +1,17 @@
+version: ad6f93f7c882bfc45c208a477c0df23ca47afe72
+type: git_submodule
+homepage: https://github.com/textmate/rez.tmbundle.git
+license: other
```
```suggestion
license: permissive
```

lildude reviewed (approved):
I've rectified the last of the issues and all looks good now. Usage appears high enough for inclusion too 👍
